ledger/simulation/simulation_eval_test.go

```diff
@@ -848,6 +848,206 @@ int 0
+			ApplicationID: futureAppID,
+			Accounts:      []basics.Address{receiver.Addr},
+		})
+
+		txntest.Group(&createTxn, &expensiveTxn)
+
+		signedCreateTxn := createTxn.Txn().Sign(sender.Sk)
+		signedExpensiveTxn := expensiveTxn.Txn().Sign(sender.Sk)
+		// Add a small bit of extra budget, but not enough
+		extraBudget := uint64(5)
+
+		return simulationTestCase{
+			input: simulation.Request{
+				TxnGroups: [][]transactions.SignedTxn{
+					{signedCreateTxn, signedExpensiveTxn},
+				},
+				ExtraOpcodeBudget: extraBudget,
+			},
+			expectedError: "dynamic cost budget exceeded",
+			expected: simulation.Result{
+				Version:   simulation.ResultLatestVersion,
+				LastRound: txnInfo.LatestRound(),
+				TxnGroups: []simulation.TxnGroupResult{
+					{
+						Txns: []simulation.TxnResult{
+							{
+								Txn: transactions.SignedTxnWithAD{
+									ApplyData: transactions.ApplyData{
+										ApplicationID: futureAppID,
+									},
+								},
+								AppBudgetConsumed: 4,
+							},
+							{
+								AppBudgetConsumed: 1401,
+							},
+						},
+						FailedAt:          simulation.TxnPath{1},
+						AppBudgetAdded:    1405,
+						AppBudgetConsumed: 1405,
+					},
+				},
+				EvalOverrides: simulation.ResultEvalOverrides{ExtraOpcodeBudget: extraBudget},
+			},
+		}
+	})
+}
+
+func TestAppCallWithExtraBudgetExceedsInternalLimit(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// Transaction group has a cost of 4 + 1404
+	expensiveAppSource := `#pragma version 6
+	txn ApplicationID      // [appId]
+	bz end                 // []
+` + strings.Repeat(`int 1; pop;`, 700) + `end:
+	int 1`
+
+	l, accounts, txnInfo := simulationtesting.PrepareSimulatorTest(t)
+	defer l.Close()
+	s := simulation.MakeSimulator(l)
+
+	sender := accounts[0]
+	receiver := accounts[1]
+
+	futureAppID := basics.AppIndex(1)
+	// App create with cost 4
+	createTxn := txnInfo.NewTxn(txntest.Txn{
+		Type:              protocol.ApplicationCallTx,
+		Sender:            sender.Addr,
+		ApplicationID:     0,
+		ApprovalProgram:   expensiveAppSource,
+		ClearStateProgram: `#pragma version 6; int 0`,
+	})
+	// Expensive 700 repetition of int 1 and pop total cost 1404
+	expensiveTxn := txnInfo.NewTxn(txntest.Txn{
+		Type:          protocol.ApplicationCallTx,
+		Sender:        sender.Addr,
+		ApplicationID: futureAppID,
+		Accounts:      []basics.Address{receiver.Addr},
+	})
+
+	txntest.Group(&createTxn, &expensiveTxn)
+
+	signedCreateTxn := createTxn.Txn().Sign(sender.Sk)
+	signedExpensiveTxn := expensiveTxn.Txn().Sign(sender.Sk)
+	// Add a small bit of extra budget, but not enough
+	extraBudget := simulation.MaxExtraOpcodeBudget + 1
+
+	// should error on too high extra budgets
+	_, err := s.Simulate(
+		simulation.Request{
+			TxnGroups:         [][]transactions.SignedTxn{{signedCreateTxn, signedExpensiveTxn}},
+			ExtraOpcodeBudget: extraBudget,
+		})
+	require.ErrorContains(t, err, "extra budget 320001 > simulation extra budget limit 320000")
```

Comment:
Could you use `require.ErrorAs` to verify that the error is an `InvalidRequestError`?

Reply:
right, I should have added this line.  See [more test in e2e for --allow-extra-opcode-budget, pr comment response](https://github.com/algorand/go-algorand/pull/5354/commits/2eebb810ca2052d22315fee59faa5f8aadd89940)